src/clp_ffi_py/ir/native/PyDeserializer.hpp

```diff
@@ -0,0 +1,255 @@
+        ) -> clp::ffi::ir_stream::IRErrorCode {
+            return m_utc_offset_change_handle(utc_offset_old, utc_offset_new);
+        }
+
+        [[nodiscard]] auto handle_schema_tree_node_insertion(
+                clp::ffi::SchemaTree::NodeLocator schema_tree_node_locator
+        ) -> clp::ffi::ir_stream::IRErrorCode {
+            return m_schema_tree_node_insertion_handle(schema_tree_node_locator);
+        }
+
+        [[nodiscard]] auto handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+            return m_end_of_stream_handle();
+        }
+
+    private:
+        // Variables
+        LogEventHandle m_log_event_handle;
+        UtcOffsetChangeHandle m_utc_offset_change_handle;
+        SchemaTreeNodeInsertionHandle m_schema_tree_node_insertion_handle;
+        EndOfStreamHandle m_end_of_stream_handle;
+    };
+
+    using Deserializer = clp::ffi::ir_stream::Deserializer<IrUnitHandler>;
+
+    // Methods
+    /**
+     * Implements `IrUnitHandler::EndOfStreamHandle`.
+     * This handle function sets the underlying `m_end_of_stream_reached` to true.
+     * @return IRErrorCode::IRErrorCode_Success on success.
+     */
+    [[maybe_unused]] auto handle_end_of_stream() -> clp::ffi::ir_stream::IRErrorCode {
+        m_end_of_stream_reached = true;
+        return clp::ffi::ir_stream::IRErrorCode::IRErrorCode_Success;
+    }
+
+    /**
+     * Implements `IrUnitHandler::LogEventHandle`.
+     * This handle function sets the underlying `m_deserialized_log_event` with the given input.
+     * @param kv_log_event
+     * @return IRErrorCode::IRErrorCode_Success on success.
+     *
+     */
+    [[nodiscard]] auto handle_log_event(clp::ffi::KeyValuePairLogEvent&& log_event
+    ) -> clp::ffi::ir_stream::IRErrorCode;
+
+    /**
+     * @return Whether `m_deserialized_log_event` has been set.
+     */
+    [[nodiscard]] auto has_unreleased_deserialized_log_event() const -> bool {
+        return nullptr != m_deserialized_log_event;
+    }
+
+    /**
+     * Releases the ownership of the underlying deserialized log event as a rvalue reference.
+     * NOTE: this method doesn't check whether the ownership is empty (nullptr). The caller must
+     * ensure the ownership is legal.
+     * @return The released ownership of the deserialized log event.
+     */
+    [[maybe_unused]] auto release_deserialized_log_event() -> clp::ffi::KeyValuePairLogEvent {
+        auto released{std::move(*m_deserialized_log_event)};
+        delete m_deserialized_log_event;
+        m_deserialized_log_event = nullptr;
+        return released;
+    }
+
+    /**
+     * Handles the incomplete IR error returned from `Deserializer::deserialize_next_ir_unit`.
+     * @param err
+     * @return true if the error is resolved on success.
+     * @return false with the relevant Python exception and error set on failures:
+     * - The error doesn't indicate an incomplete IR stream.
+     * - Incomplete IR stream is not allowed.
```

Comment:
What do these two sentences mean?
it's not very clear to me, can you elaborate?

Reply:
This function is added to "swallow" incomplete stream error if needed.

- `The error doesn't indicate an incomplete IR stream.`: Only one particular `std::error_code` indicates an incomplete stream error. If the given error isn't an incomplete stream error, we should return false as we can't handle this error.
- `Incomplete IR stream is not allowed.`: If the error code does indicate an incomplete stream error, but it is not allowed in the current deserializer, we should also return false as this is not the error we can swallow.